From 1c611eae5ac56b2a68912c616c65d424b3693fb5 Mon Sep 17 00:00:00 2001
From: Konrad Kleine <kkleine@redhat.com>
Date: Wed, 1 Mar 2023 22:32:26 +0100
Subject: [PATCH] Subpackage <package>-<toolchain>-raw-pgo-profdata

Disable by specifying this in your specfile:

```
%global _toolchain_profile_subpackages %{nil}
```

For inspiration see also:

https://src.fedoraproject.org/fork/tstellar/rpms/redhat-rpm-config/c/a62cbcf1d23df6cd2111b7cfa6f75a849d89f11d?branch=pgo
---
 macros                 | 27 ++++++++++++++++++++++++++-
 redhat-rpm-config.spec |  5 ++++-
 2 files changed, 30 insertions(+), 2 deletions(-)

diff --git a/macros b/macros
index d10fa49..4c43f18 100644
--- a/macros
+++ b/macros
@@ -26,6 +26,11 @@
 %_debugsource_packages  1
 %_debuginfo_subpackages 1
 
+# To disable generation of profile subpackage, specify this in specfile:
+# %global _toolchain_profile_subpackages %{nil}
+# TODO: Make sure to only enable it when building with clang.
+%_toolchain_profile_subpackages 1
+
 # GCC toolchain
 %__cc_gcc gcc
 %__cxx_gcc g++
@@ -49,6 +54,18 @@
 %build_cxx %{__cxx}
 %build_cpp %{__cpp}
 
+# Generate profiledata packages for the compiler
+%_toolchain_profile_subpackage_template \
+%package -n %{name}-%{toolchain}-pgo-profdata \
+Summary: %{toolchain} indexed PGO profile data from %{name} package \
+%description -n %{name}-%{toolchain}-pgo-profdata \
+This package contains profiledata for %{toolchain} that was generated while \
+compiling %{name}. This can be used for doing Profile Guided Optimizations \
+(PGO) builds of %{toolchain} \
+%files -n %{name}-%{toolchain}-pgo-profdata \
+%{_libdir}/%{toolchain}-pgo-profdata/%{name}/%{name}.%{toolchain}.profdata \
+%{nil}
+
 #==============================================================================
 # ---- compiler flags.
 
@@ -81,6 +98,9 @@
 # not been set already.  RPM_OPT_FLAGS and RPM_LD_FLAGS have already
 # been set implicitly at the start of the %%build section.
 # LT_SYS_LIBRARY_PATH is used by libtool script.
+# For LLVM_PROFILE_FILE see:
+# - https://clang.llvm.org/docs/SourceBasedCodeCoverage.html#running-the-instrumented-program
+# - https://docs.teamscale.com/howto/setting-up-profiler-tga/cpp/#clang-and-llvm-cov
 %set_build_flags \
   CFLAGS="${CFLAGS:-%{build_cflags}}" ; export CFLAGS ; \
   CXXFLAGS="${CXXFLAGS:-%{build_cxxflags}}" ; export CXXFLAGS ; \
@@ -89,7 +109,9 @@
   LDFLAGS="${LDFLAGS:-%{build_ldflags}}" ; export LDFLAGS ; \
   LT_SYS_LIBRARY_PATH="${LT_SYS_LIBRARY_PATH:-%_libdir:}" ; export LT_SYS_LIBRARY_PATH ; \
   CC="${CC:-%{__cc}}" ; export CC ; \
-  CXX="${CXX:-%{__cxx}}" ; export CXX
+  CXX="${CXX:-%{__cxx}}" ; export CXX ; \
+  [ "%_toolchain_profile_subpackages" = 1 ] && TMPDIR="%{_builddir}/raw-pgo-profdata" && export TMPDIR && mkdir -pv $TMPDIR && \
+  LLVM_PROFILE_FILE="%t/%{name}.%{toolchain}.%m.profraw" && export LLVM_PROFILE_FILE ;
 
 # Automatically use set_build_flags macro for build, check, and
 # install phases.
@@ -253,6 +275,7 @@ print(result)
 %__brp_mangle_shebangs /usr/lib/rpm/redhat/brp-mangle-shebangs %{?__brp_mangle_shebangs_exclude:--shebangs "%{?__brp_mangle_shebangs_exclude}"} %{?__brp_mangle_shebangs_exclude_file:--shebangs-from "%{__brp_mangle_shebangs_exclude_file}"} %{?__brp_mangle_shebangs_exclude_from:--files "%{?__brp_mangle_shebangs_exclude_from}"} %{?__brp_mangle_shebangs_exclude_from_file:--files-from "%{__brp_mangle_shebangs_exclude_from_file}"}
 
 %__brp_llvm_compile_lto_elf /usr/lib/rpm/redhat/brp-llvm-compile-lto-elf %{build_cflags} %{build_ldflags}
+%__copy_profraw %[ 0%{_toolchain_profile_subpackages} > 0 ? "mkdir -pv %{buildroot}%{_libdir}/%{toolchain}-pgo-profdata/%{name} && find %{_builddir}/raw-pgo-profdata -type f -name '%{name}.%{toolchain}.*.profraw' > %{_builddir}/pgo-profiles && llvm-profdata merge --debug-info-correlate --enable-name-compression -sparse $(cat %{_builddir}/pgo-profiles) -o %{buildroot}%{_libdir}/%{toolchain}-pgo-profdata/%{name}/%{name}.%{toolchain}.profdata" : "%{nil}" ]
 
 # note: %%__os_install_post_python is defined in python-srpm-macros and contains several policies
 # redhat-rpm-config maintainers, don't remove it from %%__os_install_post unless coordinating the change with Python maintainers
@@ -271,6 +294,7 @@ print(result)
     %{?__brp_mangle_shebangs} \
     %{?__brp_remove_la_files} \
     %{__os_install_post_python} \
+    %{?_toolchain_profile_subpackages:%{?__copy_profraw}} \
 %{nil}
 
 %__spec_install_post\
@@ -281,6 +305,7 @@ print(result)
 %{nil}
 
 %install %{?_enable_debug_packages:%{?buildsubdir:%{debug_package}}}\
+         %[ 0%{_toolchain_profile_subpackages} > 0 ? "%{_toolchain_profile_subpackage_template}" : "%{nil}" ] \
 %%install\
 %{nil}
 
diff --git a/redhat-rpm-config.spec b/redhat-rpm-config.spec
index 09a5b7f..694b900 100644
--- a/redhat-rpm-config.spec
+++ b/redhat-rpm-config.spec
@@ -4,7 +4,7 @@
 # 2) When making changes, increment the version (in baserelease) by 1.
 #    rpmdev-bumpspec and other tools update the macro below, which is used
 #    in Version: to get the desired effect.
-%global baserelease 229
+%global baserelease 230
 
 Summary: Red Hat specific rpm configuration files
 Name: redhat-rpm-config
@@ -222,6 +222,9 @@ install -p -m 644 -t %{buildroot}%{_rpmluadir}/fedora/srpm forge.lua
 %doc buildflags.md
 
 %changelog
+* Wed Mar 01 2023 Konrad Kleine <kkleine@redhat.com> - 230-1
+- Automatically add <PACKAGE>-<toolchain>-profdata subpackage (only for clang now)
+
 * Tue Jan 17 2023 Miro Hronƒçok <mhroncok@redhat.com> - 229-1
 - Add pyproject-srpm-macros to the default buildroot
 
-- 
2.39.1

